mathematica/retrobop_dynamical_stability.cpp

```diff
@@ -0,0 +1,668 @@
+  }
+  int total_breakdowns = 0;
+  // If the error between an integration at step and one at step/2 exceeds this
+  // over a year, we assume that things have happened that our integrator cannot
+  // handle, probably close encounters.
+  // TODO(egg): This is a very lousy substitute for a proper estimation of the
+  // local forward error.  We probably want to have a way to actually estimate
+  // the local error (on every step), and perhaps even the local backward error
+  // (though that may be costly if done naïvely).
+  Length const yearly_allowed_numerical_error = 1000 * Metre;
+
+  for (int year = 1; year <= 200; ++year) {
+    Instant const t = ksp_epoch + year * JulianYear;
+    Bundle bundle{7};
+    for (auto const& ephemeris : perturbed_ephemerides) {
+      bundle.Add([
+        &numerically_unsound,
+        ephemeris = ephemeris.get(),
+        t,
+        yearly_allowed_numerical_error
+      ]() {
+        auto system = ReadSystem(system_file);
+        for (auto const celestial : celestials) {
+          system.degrees_of_freedom[celestial] =
+              EvaluateDegreesOfFreedom(*ephemeris,
+                                       celestial,
+                                       ephemeris->t_min());
+        }
+        Ephemeris<Barycentric> refined(
+            std::move(system.bodies),
+            system.degrees_of_freedom,
+            ephemeris->t_min(),
+            1 * Milli(Metre),
+            Ephemeris<Barycentric>::FixedStepParameters(
+                integrators::QuinlanTremaine1990Order12<
+                    Position<Barycentric>>(),
+                step / 2));
+        ephemeris->Prolong(t);
+        ephemeris->ForgetBefore(t);
+        refined.Prolong(t);
+        Length numerical_error;
+        Celestial most_erroneous_moon;
+        ComputeHighestMoonError(refined,
+                                *ephemeris,
+                                t,
+                                numerical_error,
+                                most_erroneous_moon);
+        if (numerical_error > yearly_allowed_numerical_error) {
+          LOG(INFO) << "high numerical error " << numerical_error << " ("
+                    << names[most_erroneous_moon] << ")";
+          numerically_unsound[ephemeris] = true;
+        }
+        return Status::OK;
+      });
+    }
+    bundle.Join();
+    LOG(INFO) << "year " << year;
+
+    int yearly_breakdowns = 0;
+    for (auto it = perturbed_ephemerides.begin();
+         it != perturbed_ephemerides.end();) {
+      if (numerically_unsound[it->get()]) {
+        goto erase_and_skip_to_next_perturbed_ephemeris;
+      } else {
+        Ephemeris<Barycentric> const& ephemeris = **it;
+        auto const jool_barycentre =
+            JoolSystemBarycentre(ephemeris, t).position();
+        for (auto const moon : jool_moons) {
+          Length const distance =
+              (EvaluatePosition(ephemeris, moon, t) - jool_barycentre).Norm();
+          if (distance > 3e8 * Metre) {
```

Comment:
This kind of threshold (there are others in this code) should be made constants at the beginning of the file.